Automattic/isolated-block-editor
# Collab: Fix handling of initial content from `onLoad`

After testing against an actual remote server, I found that a timing bug could manifest in cases where the initial iso-editor content is loaded in an `onLoad` handler. When this happens it will wipe out the editor content whenever a peer connects.

This fix ensures that the Yjs doc is always initialized with up-to-date blocks.

<details>
<summary>Bug details</summary>
<p>

Order of events:

1. iso-editor first renders with `blocks = null`
2. `onLoad` runs and dispatches an action to set Gutenberg data to `blocks = [ { ... } ]`
3. `initYDoc` requests a server connection, and has a reference to `blocks` which is still null
4. a re-render is triggered as a result of step 2
5. `initYDoc` successfully connects but still thinks `blocks = null` and initializes its Yjs data with `blocks = null`
6. a peer client joins the channel and requests data from the first client
7. the first client sends Yjs data with `blocks = null`
8. the second client receives that Yjs data and updates Gutenberg data to `blocks = null`
9. that update is sent back to the first client
10. the first client now also updates its Gutenberg data to `blocks = null`

When there is zero server lag (i.e. responds synchronously), the order of 4 and 5 is flipped and doesn't cause a problem.

</p>
</details>

## Testing instructions

1. `yarn storybook`
2. Load http://localhost:6006/?path=/story/collaboration--with-on-load in one window.
3. Open the same URL in another window.

The initial content should not be wiped out.
